README.rst
Custom Clock for LCD 16x02 (based on RPLCD-i2c)
#####

Thanks to XXXX for the RPLCD-i2c driver.
https://github.com/zador-blood-stained/RPLCD-i2c

A Python 2/3 Raspberry Pi Custom Clock for the Hitachi HD44780 LCD

Tested with the 16x2 LCD MT-20S4A??, Raspberry Pi3 and PCF8574AT
I²C port expander.

Original GPIO based library tested with:
- 20x4 LCD that is sold for example by adafruit.com or mikroshop.ch
- 16x2 LCD from mikroshop.ch

Depends on `python-smbus` library.
For getting smbus support in Python 3.x use instructions provided here:

http://procrastinative.ninja/2014/07/21/smbus-for-python34-on-raspberry/

http://jtecheng.com/?p=959

Wiring
========

Refer to original github repo for more details
https://github.com/zador-blood-stained/RPLCD-i2c

Installation
========

.. code::

     git clone https://github.com/bousqi/Clock_RPLCD-i2c.git
     cd Clock_RPLCD-i2c
     sudo python3 setup.py install
     # and/or
     sudo python setup.py install


I²C bus number
-----------

For Rasbberry Pi Model A, B Rev 2, B+, Raspberry Pi 2 and Raspberry Pi 3 use bus number 1
For Raspberry Pi Model B Rev 1 use bus number 0

I²C device address
-----------

To check your port expander address use ``gpio i2cd`` command
(alternatively ``sudo i2cdetect -y 0``
or ``sudo i2cdetect -y 1`` depending on bus number)
Example output with one PCF8574AT connected at address 0x3F:

.. code::

     % sudo i2cdetect -y 1
          0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
     00:          -- -- -- -- -- -- -- -- -- -- -- -- --
     10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
     20: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
     30: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 3F
     40: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
     50: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
     60: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
     70: -- -- -- -- -- -- -- --
     %

or refer to datasheet for your port expander.

Features, Code examples & Usage
========

blabla

Testing
=======

To test your 20x4 display, please run the ``test_20x4.py`` script and
confirm/verify each step with the enter key. Make sure to set your bus 
number and address to the ``CharLCD`` constructor in ``test_20x4.py``.

To test your 16x2 display, please run the ``test_16x2.py`` script and
confirm/verify each step with the enter key. Make sure to set your bus 
number and address to the ``CharLCD`` constructor in ``test_16x2.py``.

You can check the charmap on your display with ``show_charmap.py 2 16``
on a 16x2 display.

License
=======

This code is licensed under the MIT license, see the `LICENSE file
<https://github.com/zador-blood-stained/RPLCD-i2c/blob/master/LICENSE>`_ or `tldrlegal
<http://www.tldrlegal.com/license/mit-license>`_ for more information. 

The module ``RPLCD/enum.py`` is (c) 2004-2013 by Barry Warsaw. It was
distributed as part of the ``flufl.enum`` package under the LGPL License version
3 or later.


CHANGELOG
v0.5.0

- Initial release


test_16x2.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
from __future__ import print_function, division, absolute_import, unicode_literals

import sys

from RPLCD_i2c import CharLCD
from RPLCD_i2c import Alignment, CursorMode, ShiftMode
from RPLCD_i2c import cursor, cleared

try:
    input = raw_input
except NameError:
    pass

try:
    unichr = unichr
except NameError:
    unichr = chr


lcd = CharLCD(address=0x3F, port=1, cols=16, rows=2, dotsize=8)

input('Display should be blank. ')

lcd.cursor_mode = CursorMode.blink
input('The cursor should now blink. ')

lcd.cursor_mode = CursorMode.line
input('The cursor should now be a line. ')

lcd.write_string('Hello world!')
input('"Hello world!" should be on the LCD. ')

assert lcd.cursor_pos == (0, 12), 'cursor_pos should now be (0, 12)'

lcd.cursor_pos = (1, 0)
lcd.write_string('2')
assert lcd.cursor_pos == (1, 1), 'cursor_pos should now be (2, 1)'
input('Lines 2, should now be labelled with the right number. ')

lcd.clear()
input('Display should now be clear, cursor should be at initial position. ')

lcd.cursor_pos = (0, 5)
lcd.write_string('12345')
input('The string should have a left offset of 5 characters. ')

lcd.write_shift_mode = ShiftMode.display
lcd.cursor_pos = (1, 5)
lcd.write_string('12345')
input('Both strings should now be at column 0. ')

lcd.write_shift_mode = ShiftMode.cursor
lcd.cursor_pos = (0, 5)
lcd.write_string(lcd.write_shift_mode.name)
input('The string "cursor" should now be on the first row, column 0. ')

lcd.home()
input('Cursor should now be at initial position. Everything should be shifted to the right by 5 characters. ')

with cursor(lcd, 1, 15):
    lcd.write_string('X')
input('The last character on the LCD should now be an "X"')

lcd.display_enabled = False
input('Display should now be blank. ')

with cleared(lcd):
    lcd.write_string('Eggs Ham Bacon\n\rand Spam')
lcd.display_enabled = True
input('Display should now show "Eggs Ham Bacon and Spam". ')

lcd.shift_display(4)
input('Text should now be shifted to the right by 4 characters. ')
lcd.shift_display(-4)
input('Shift should now be undone. ')

lcd.text_align_mode = Alignment.right
lcd.cursor_mode = CursorMode.hide
lcd.write_string(' Spam')
input('The word "Spam" should now be inverted. ')

lcd.text_align_mode = Alignment.left
lcd.cursor_mode = CursorMode.hide
lcd.write_string(' Wurscht')
input('The word "mapS" should now be replaced with "Wurscht". ')

lcd.clear()
lcd.write_string('1\n')
lcd.write_string('2')
input('The numbers 1-2 should now be displayed, each line shifted to the right by 1 char more than the previous. ')

lcd.clear()
lcd.write_string('This is a multiple line string!')
input('Text should nicely wrap around lines. ')

lcd.cursor_mode = CursorMode.hide

# Test custom chars
lcd.clear()
happy = (0b00000, 0b01010, 0b01010, 0b00000, 0b10001, 0b10001, 0b01110, 0b00000)
sad = (0b00000, 0b01010, 0b01010, 0b00000, 0b01110, 0b10001, 0b10001, 0b00000)
lcd.create_char(0, sad)
lcd.write_string(unichr(0))
lcd.create_char(1, happy)
lcd.write_string(unichr(1))
input('You should now see a sad and a happy face next to each other. ')
lcd.create_char(0, happy)
lcd.home()
lcd.write_string(unichr(0))
input('Now both faces should be happy. ')

lcd.clear()
lcd.set_backlight(False)
lcd.home()
lcd.write_string('No backlight')
input('Display backlight should be off (if wired). ')

lcd.clear()
lcd.set_backlight(True)
lcd.home()
lcd.write_string('Backlight')
input('Display backlight should be back on (if wired). ')

lcd.clear()
print('Test done.')


test_20x4.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
from __future__ import print_function, division, absolute_import, unicode_literals

import sys

from RPLCD_i2c import CharLCD
from RPLCD_i2c import Alignment, CursorMode, ShiftMode
from RPLCD_i2c import cursor, cleared

try:
    input = raw_input
except NameError:
    pass

try:
    unichr = unichr
except NameError:
    unichr = chr


lcd = CharLCD(address=0x38, port=1, cols=20, rows=4, dotsize=8)

input('Display should be blank. ')

lcd.cursor_mode = CursorMode.blink
input('The cursor should now blink. ')

lcd.cursor_mode = CursorMode.line
input('The cursor should now be a line. ')

lcd.write_string('Hello world!')
input('"Hello world!" should be on the LCD. ')

assert lcd.cursor_pos == (0, 12), 'cursor_pos should now be (0, 12)'

lcd.cursor_pos = (1, 0)
lcd.write_string('2')
lcd.cursor_pos = (2, 0)
lcd.write_string('3')
lcd.cursor_pos = (3, 0)
lcd.write_string('4')
assert lcd.cursor_pos == (3, 1), 'cursor_pos should now be (3, 1)'
input('Lines 2, 3 and 4 should now be labelled with the right numbers. ')

lcd.clear()
input('Display should now be clear, cursor should be at initial position. ')

lcd.cursor_pos = (0, 5)
lcd.write_string('12345')
input('The string should have a left offset of 5 characters. ')

lcd.write_shift_mode = ShiftMode.display
lcd.cursor_pos = (1, 5)
lcd.write_string('12345')
input('Both strings should now be at column 0. ')

lcd.write_shift_mode = ShiftMode.cursor
lcd.cursor_pos = (2, 5)
lcd.write_string(lcd.write_shift_mode.name)
input('The string "cursor" should now be on the third row, column 0. ')

lcd.home()
input('Cursor should now be at initial position. Everything should be shifted to the right by 5 characters. ')

with cursor(lcd, 3, 19):
    lcd.write_string('X')
input('The last character on the LCD should now be an "X"')

lcd.display_enabled = False
input('Display should now be blank. ')

with cleared(lcd):
    lcd.write_string('Eggs, Ham, Bacon\n\rand Spam')
lcd.display_enabled = True
input('Display should now show "Eggs, Ham, Bacon and Spam". ')

lcd.shift_display(4)
input('Text should now be shifted to the right by 4 characters. ')
lcd.shift_display(-4)
input('Shift should now be undone. ')

lcd.text_align_mode = Alignment.right
lcd.cursor_mode = CursorMode.hide
lcd.write_string(' Spam')
input('The word "Spam" should now be inverted. ')

lcd.text_align_mode = Alignment.left
lcd.cursor_mode = CursorMode.hide
lcd.write_string(' Wurscht')
input('The word "mapS" should now be replaced with "Wurscht". ')

lcd.clear()
lcd.write_string('1\n')
lcd.write_string('2\n')
lcd.write_string('3\n')
lcd.write_string('4')
input('The numbers 1-4 should now be displayed, each line shifted to the right by 1 char more than the previous. ')

lcd.clear()
lcd.write_string('This is a long string that will wrap across multiple lines!')
input('Text should nicely wrap around lines. ')

lcd.cursor_mode = CursorMode.hide

# Test custom chars
lcd.clear()
happy = (0b00000, 0b01010, 0b01010, 0b00000, 0b10001, 0b10001, 0b01110, 0b00000)
sad = (0b00000, 0b01010, 0b01010, 0b00000, 0b01110, 0b10001, 0b10001, 0b00000)
lcd.create_char(0, sad)
lcd.write_string(unichr(0))
lcd.create_char(1, happy)
lcd.write_string(unichr(1))
input('You should now see a sad and a happy face next to each other. ')
lcd.create_char(0, happy)
lcd.home()
lcd.write_string(unichr(0))
input('Now both faces should be happy. ')

lcd.clear()
lcd.set_backlight(False)
lcd.home()
lcd.write_string('No backlight')
input('Display backlight should be off (if wired). ')

lcd.clear()
lcd.set_backlight(True)
lcd.home()
lcd.write_string('Backlight')
input('Display backlight should be back on (if wired). ')

lcd.clear()
print('Test done.')


show_charmap.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
Copyright (C) 2013-2015 Danilo Bargen

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
the Software, and to permit persons to whom the Software is furnished to do so,
subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

"""
from __future__ import print_function, division, absolute_import, unicode_literals

import sys

from RPLCD_i2c import CharLCD


try:
    range = xrange
except NameError:  # Python 3
    pass

try:
    safe_input = raw_input
except NameError:  # Python 3
    safe_input = input

try:
    unichr = unichr
except NameError:  # Python 3
    unichr = chr


if __name__ == '__main__':
    if len(sys.argv) != 3:
        print('Usage: %s <rows> <cols>' % sys.argv[0])
        sys.exit(1)

    rows, cols = int(sys.argv[1]), int(sys.argv[2])
    lcd = CharLCD(address=0x3F, port=1, cols=cols, rows=rows, dotsize=8, ignore_special=True)

    print('This tool shows the character map of your LCD on the display.')
    print('Press ctrl+c at any time to abort.\n')

    page = 0
    chars_per_page = rows * cols

    try:
        while True:
            page_start = page * chars_per_page
            page_end = page_start + chars_per_page
            if page_end > 256:
                page_end = 256
            lcd.clear()
            print('Displaying page %d (characters %d-%d).' %
                       (page, page_start, page_end - 1))
            for i in range(page_start, page_end):
                if i < 255:
                    lcd.write_string(unichr(i))
                else:
                    lcd.write_string(unichr(i))
                    safe_input('Press <ENTER> to exit.')
                    lcd.clear()
                    sys.exit(0)
            page += 1
            safe_input('Press <ENTER> to continue.')
    except KeyboardInterrupt:
        print('Aborting.')

    lcd.clear()


clock.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
from __future__ import print_function, division, absolute_import, unicode_literals

import sys
import bme280

from RPLCD_i2c import CharLCD
from RPLCD_i2c import Alignment, CursorMode, ShiftMode
from RPLCD_i2c import cursor, cleared
from time import strftime, sleep
from datetime import datetime

# build big digits
def disp_number(lcd_char, position):

    if lcd_char=="0" :
        lcd.cursor_pos = (0, position)
        lcd.write_string(unichr(4))
        lcd.write_string(unichr(0))
        lcd.write_string(unichr(5))
        lcd.cursor_pos = (1, position)
        lcd.write_string(unichr(7))
        lcd.write_string(unichr(1))
        lcd.write_string(unichr(6))

    elif lcd_char=="1" :
        lcd.cursor_pos = (0, position)
        lcd.write_string(unichr(0))
        lcd.write_string(unichr(5))
        lcd.write_string(unichr(254))
        lcd.cursor_pos = (1, position)
        lcd.write_string(unichr(1))
        lcd.write_string(unichr(255))
        lcd.write_string(unichr(1))
		
    elif lcd_char=="2" :
        lcd.cursor_pos = (0, position)
        lcd.write_string(unichr(2))
        lcd.write_string(unichr(2))
        lcd.write_string(unichr(5))
        lcd.cursor_pos = (1, position)
        lcd.write_string(unichr(255))
        lcd.write_string(unichr(1))
        lcd.write_string(unichr(1))

    elif lcd_char=="3" :
        lcd.cursor_pos = (0, position)
        lcd.write_string(unichr(0))
        lcd.write_string(unichr(2))
        lcd.write_string(unichr(5))
        lcd.cursor_pos = (1, position)
        lcd.write_string(unichr(1))
        lcd.write_string(unichr(1))
        lcd.write_string(unichr(6))

    elif lcd_char=="4" :
        lcd.cursor_pos = (0, position)
        lcd.write_string(unichr(255))
        lcd.write_string(unichr(1))
        lcd.write_string(unichr(255))
        lcd.cursor_pos = (1, position)
        lcd.write_string(unichr(254))
        lcd.write_string(unichr(254))
        lcd.write_string(unichr(6))
		
    elif lcd_char=="5" :
        lcd.cursor_pos = (0, position)
        lcd.write_string(unichr(255))
        lcd.write_string(unichr(2))
        lcd.write_string(unichr(2))
        lcd.cursor_pos = (1, position)
        lcd.write_string(unichr(1))
        lcd.write_string(unichr(1))
        lcd.write_string(unichr(6))
		
    elif lcd_char=="6" :
        lcd.cursor_pos = (0, position)
        lcd.write_string(unichr(4))
        lcd.write_string(unichr(2))
        lcd.write_string(unichr(2))
        lcd.cursor_pos = (1, position)
        lcd.write_string(unichr(7))
        lcd.write_string(unichr(1))
        lcd.write_string(unichr(6))
		
    elif lcd_char=="7" :
        lcd.cursor_pos = (0, position)
        lcd.write_string(unichr(0))
        lcd.write_string(unichr(0))
        lcd.write_string(unichr(5))
        lcd.cursor_pos = (1, position)
        lcd.write_string(unichr(254))
        lcd.write_string(unichr(254))
        lcd.write_string(unichr(6))
		
    elif lcd_char=="8" :
        lcd.cursor_pos = (0, position)
        lcd.write_string(unichr(4))
        lcd.write_string(unichr(2))
        lcd.write_string(unichr(5))
        lcd.cursor_pos = (1, position)
        lcd.write_string(unichr(7))
        lcd.write_string(unichr(1))
        lcd.write_string(unichr(6))
		
    elif lcd_char=="9" :
        lcd.cursor_pos = (0, position)
        lcd.write_string(unichr(255))
        lcd.write_string(unichr(2))
        lcd.write_string(unichr(5))
        lcd.cursor_pos = (1, position)
        lcd.write_string(unichr(1))
        lcd.write_string(unichr(1))
        lcd.write_string(unichr(6))
     
    elif lcd_char=="C" :
        lcd.cursor_pos = (0, position)
        lcd.write_string(unichr(255))
        lcd.write_string(unichr(2))
        lcd.cursor_pos = (1, position)
        lcd.write_string(unichr(254))
        lcd.write_string(unichr(254))
    
    else:
        lcd.cursor_pos = (0, position)
        lcd.write_string(unichr(254))
        lcd.write_string(unichr(254))
        lcd.write_string(unichr(254))
        lcd.cursor_pos = (1, position)
        lcd.write_string(unichr(254))
        lcd.write_string(unichr(254))
        lcd.write_string(unichr(254))
    
    return lcd_char;
    
# scroll effect
def shift(direction): 
    if direction=="left" :
        for x in range(0, 16):
            lcd.shift_display(-1)
            sleep(0.05)
    else :
        for x in range(0, 16):
            lcd.shift_display(1)
            sleep(0.05)

def read_temp():
    temperature,pressure,humidity = bme280.readBME280All()
    return temperature
                
def clock_dots():
    # display two dots
    lcd.cursor_pos = (0, 8)
    lcd.write_string(unichr(3))
    lcd.cursor_pos = (1, 8)
    lcd.write_string(unichr(3))
    sleep(0.5)
    
    # remove the two dots
    lcd.cursor_pos = (0, 8)
    lcd.write_string(unichr(254))
    lcd.cursor_pos = (1, 8)
    lcd.write_string(unichr(254))
    sleep(0.5)

def clock_date(digits, month, day_name):
    # previous text leaving to left
    shift("left")
    # preparing date buffer
    tens_day = disp_number(digits[4], 0)
    day = disp_number(digits[5], 4)
    lcd.cursor_pos = (0, 8)
    lcd.write_string('        ')
    lcd.cursor_pos = (1, 8)
    lcd.write_string('        ')
    lcd.cursor_pos = (0, 8)
    lcd.write_string(month[:8])
    lcd.cursor_pos = (1, 8)
    lcd.write_string(day_name[:8])
    # date entering from left
    shift("right")
    sleep(2)


def clock_temp():
    # previous text leaving to left
    shift("left")
    # preparing temp buffer
    lcd.cursor_pos = (0, 0)
    lcd.write_string('Temp:           ')
    lcd.cursor_pos = (1, 0)
    lcd.write_string('                ')

    temp_digits = str(read_temp());
    disp_number(temp_digits[0], 6)
    disp_number(temp_digits[1], 10)
    disp_number("C", 14)
    lcd.cursor_pos = (0, 13)
    lcd.write_string(unichr(223))
    # date entering from left
    shift("right")
    sleep(5)

def clock_hour(digits):
    # date leaving to left
    shift("left")
    # preparing hour buffer
    lcd.cursor_pos = (0, 0)
    lcd.write_string('                ')
    lcd.cursor_pos = (1, 0)
    lcd.write_string('                ')
    disp_number(digits[0], 0)
    disp_number(digits[1], 4)
    disp_number(digits[2], 9)
    disp_number(digits[3], 13)
    # hour entering from left
    shift("right")


def main():

    try:
        input = raw_input
    except NameError:
        pass

    try:
        unichr = unichr
    except NameError:
        unichr = chr

    old_time = 0
    counter = 0

    # custom symbols
    lcd.clear()
    top_line = (0b11111, 0b11111, 0b00000, 0b00000, 0b00000, 0b00000, 0b00000, 0b00000)
    bottom_line = (0b00000, 0b00000, 0b00000, 0b00000, 0b00000, 0b00000, 0b11111, 0b11111)
    both_lines = (0b11111, 0b11111, 0b00000, 0b00000, 0b00000, 0b00000, 0b11111, 0b11111)
    dot = (0b00000, 0b00000, 0b00000, 0b11000, 0b11000, 0b00000, 0b00000, 0b00000)
    up_left = (0b00111, 0b01111, 0b01111, 0b11111, 0b11111, 0b11111, 0b11111, 0b11111)
    up_right = (0b11100, 0b11110, 0b11110, 0b11111, 0b11111, 0b11111, 0b11111, 0b11111)
    bot_right = (0b11111, 0b11111, 0b11111, 0b11111, 0b11111, 0b11110, 0b11110, 0b11100)
    bot_left = (0b11111, 0b11111, 0b11111, 0b11111, 0b11111, 0b01111, 0b01111, 0b00111)

    lcd.create_char(0, top_line)
    lcd.create_char(1, bottom_line)
    lcd.create_char(2, both_lines)
    lcd.create_char(3, dot)
    lcd.create_char(4, up_left)
    lcd.create_char(5, up_right)
    lcd.create_char(6, bot_right)
    lcd.create_char(7, bot_left)

    #main loop
    while True:
        counter += 1
        new_time = datetime.now().strftime('%H%M%d')
        month = datetime.now().strftime('%B')
        day_name = datetime.now().strftime('%A')

        # time changed, update LCD buffer
        if new_time!=old_time :
            digits = str(new_time)
            tens_hour = disp_number(digits[0], 0)
            hour = disp_number(digits[1], 4)
            tens_minutes = disp_number(digits[2], 9)
            minutes = disp_number(digits[3], 13)
            old_time = new_time

            cur_hour = int(tens_hour)*10 + int(hour)
            # enable backlight by night
            lcd.set_backlight(cur_hour>=19 or cur_hour<9)

        # dots blink
        clock_dots()
        
        if counter==15 :
            counter=0
            # date display
            clock_date(digits, month, day_name)
            # temperature display
            clock_temp()
            # temperature display
            clock_hour(digits)


if __name__ == '__main__':

  lcd = CharLCD(address=0x3F, port=1, cols=16, rows=2, dotsize=8)

  try:
    main()
  except KeyboardInterrupt:
    pass
  finally:
    lcd.clear()
    lcd.set_backlight(False)
    lcd.home()
